File: common/src/main/java/org/valkyrienskies/clockwork/util/builder/BuilderTransformersClockwork.java
package org.valkyrienskies.clockwork.util.builder;


import com.simibubi.create.AllBlocks;
import com.simibubi.create.Create;
import com.simibubi.create.content.contraptions.base.RotatedPillarKineticBlock;
import com.simibubi.create.content.contraptions.relays.encased.EncasedCTBehaviour;
import com.simibubi.create.foundation.block.BlockStressDefaults;
import com.simibubi.create.foundation.block.connected.CTSpriteShiftEntry;
import com.simibubi.create.foundation.data.AssetLookup;
import com.simibubi.create.foundation.data.CreateRegistrate;
import com.simibubi.create.foundation.data.SharedProperties;
import com.tterrag.registrate.builders.BlockBuilder;
import com.tterrag.registrate.builders.FluidBuilder;
import com.tterrag.registrate.fabric.SimpleFlowableFluid;
import com.tterrag.registrate.util.nullness.NonNullUnaryOperator;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.level.ItemLike;
import net.minecraft.world.level.block.Block;
import net.minecraft.world.level.block.state.BlockBehaviour;
import org.valkyrienskies.clockwork.ClockWorkMod;
import org.valkyrienskies.clockwork.content.contraptions.casing.ExtendedEncasedShaftBlock;

import java.util.function.Supplier;

import static com.simibubi.create.foundation.data.BlockStateGen.axisBlock;


public class BuilderTransformersClockwork {


    private static <B extends RotatedPillarKineticBlock, P> BlockBuilder<B, P> encasedBase(BlockBuilder<B, P> b,
                                                                                           Supplier<ItemLike> drop) {
        return b.initialProperties(SharedProperties::stone)
                .properties(BlockBehaviour.Properties::noOcclusion)
                .transform(BlockStressDefaults.setNoImpact())
                .loot((p, lb) -> p.dropOther(lb, drop.get()));
    }

    public static <B extends Block, P> NonNullUnaryOperator<BlockBuilder<B, P>> infuser() {
        ResourceLocation baseBlockModelLocation = ClockWorkMod.asResource("block/physics_infuser/block");
        ResourceLocation baseItemModelLocation = ClockWorkMod.asResource("block/physics_infuser/item");
        ResourceLocation liquidTextureLocation = ClockWorkMod.asResource("block/physics_infuser/liquid");
        ResourceLocation coreTextureLocation = ClockWorkMod.asResource("block/physics_infuser/core");

        return b -> b.initialProperties(SharedProperties::stone)
                .properties(p -> p.noOcclusion())
                .blockstate((c, p) -> p.directionalBlock(c.get(), p.models()
                        .withExistingParent(c.getName(), baseBlockModelLocation)
                        .texture("0", baseBlockModelLocation)
                        .texture("1", coreTextureLocation)
                        .texture("2", liquidTextureLocation)))
                .item()
                .model((c, p) -> p.withExistingParent(c.getName(), baseItemModelLocation)
                        .texture("0", baseBlockModelLocation)
                        .texture("1", coreTextureLocation)
                        .texture("2", liquidTextureLocation))
                .build();
    }

    public static <B extends Block, P> NonNullUnaryOperator<BlockBuilder<B, P>> flapbearing() {
        ResourceLocation baseBlockModelLocation = ClockWorkMod.asResource("block/flap_bearing/block");
        ResourceLocation baseItemModelLocation = ClockWorkMod.asResource("block/bearing/item");
        ResourceLocation topTextureLocation = ClockWorkMod.asResource("block/flap_bearing/top");
        ResourceLocation baseTextureLocation = ClockWorkMod.asResource("block/flap_bearing");
        return b -> b.initialProperties(SharedProperties::stone)
                .properties(p -> p.noOcclusion())
                .blockstate((c, p) -> p.directionalBlock(c.get(), p.models()
                        .withExistingParent(c.getName(), baseBlockModelLocation)
                        .texture("0", baseBlockModelLocation)
                        .texture("1", topTextureLocation)
                        .texture("2", baseTextureLocation)))
                .item()
                .model((c, p) -> p.withExistingParent(c.getName(), baseItemModelLocation)
                        .texture("0", baseBlockModelLocation)
                        .texture("1", topTextureLocation)
                        .texture("2", baseTextureLocation))
                .build();
    }

    public static <B extends ExtendedEncasedShaftBlock, P> NonNullUnaryOperator<BlockBuilder<B, P>> encasedShaft(String casing,
                                                                                                                 Supplier<CTSpriteShiftEntry> casingShift) {
        return builder -> encasedBase(builder, () -> AllBlocks.SHAFT.get())
                .onRegister(CreateRegistrate.connectedTextures(() -> new EncasedCTBehaviour(casingShift.get())))
                .onRegister(CreateRegistrate.casingConnectivity((block, cc) -> cc.make(block, casingShift.get(),
                        (s, f) -> f.getAxis() != s.getValue(ExtendedEncasedShaftBlock.AXIS))))
                .blockstate((c, p) -> axisBlock(c, p, blockState -> p.models()
                        .getExistingFile(p.modLoc("block/encased_shaft/block_" + casing)), true))
                .item()
                .model(AssetLookup.customBlockItemModel("encased_shaft", "item_" + casing))
                .build();
    }

}

File: common/src/main/kotlin/org/valkyrienskies/clockwork/util/builder/BuilderTransformersClockwork.kt
package org.valkyrienskies.clockwork.util.builder

import com.simibubi.create.Create
import com.simibubi.create.content.kinetics.base.RotatedPillarKineticBlock
import com.simibubi.create.foundation.data.SharedProperties
import com.tterrag.registrate.builders.BlockBuilder
import com.tterrag.registrate.providers.DataGenContext
import com.tterrag.registrate.providers.RegistrateBlockstateProvider
import com.tterrag.registrate.providers.RegistrateItemModelProvider
import com.tterrag.registrate.util.nullness.NonNullBiConsumer
import com.tterrag.registrate.util.nullness.NonNullSupplier
import com.tterrag.registrate.util.nullness.NonNullUnaryOperator
import net.minecraft.resources.ResourceLocation
import net.minecraft.world.item.BlockItem
import net.minecraft.world.item.Item
import net.minecraft.world.level.ItemLike
import net.minecraft.world.level.block.Block
import net.minecraft.world.level.block.state.BlockBehaviour
import org.valkyrienskies.clockwork.ClockworkMod
import org.valkyrienskies.clockwork.ClockworkStress
import java.util.function.Supplier

object BuilderTransformersClockwork {
    private fun <B : RotatedPillarKineticBlock, P> encasedBase(
        b: BlockBuilder<B, P>,
        drop: Supplier<ItemLike>
    ): BlockBuilder<B, P> {
        return b.initialProperties { SharedProperties.stone() }
            .properties(BlockBehaviour.Properties::noOcclusion)
            .transform(ClockworkStress.setNoImpact<B, P>())
            .loot { p, lb -> p.dropOther(lb, drop.get()) }
    }

    fun <B : Block?, P> infuser(): NonNullUnaryOperator<BlockBuilder<B, P>> {
        val baseBlockModelLocation: ResourceLocation = ClockworkMod.asResource("block/physics_infuser/block")
        val baseItemModelLocation: ResourceLocation = ClockworkMod.asResource("block/physics_infuser/item")
        val liquidTextureLocation: ResourceLocation = ClockworkMod.asResource("block/physics_infuser/liquid")
        val coreTextureLocation: ResourceLocation = ClockworkMod.asResource("block/physics_infuser/core")
        return NonNullUnaryOperator { b: BlockBuilder<B, P> ->
            b.initialProperties { SharedProperties.stone() }
                .properties { p: BlockBehaviour.Properties -> p.noOcclusion() }
                .blockstate { c: DataGenContext<Block?, B>, p: RegistrateBlockstateProvider ->
                    p.directionalBlock(
                        c.get(),
                        p.models()
                            .withExistingParent(c.name, baseBlockModelLocation)
                            .texture("0", baseBlockModelLocation)
                            .texture("1", coreTextureLocation)
                            .texture("2", liquidTextureLocation)
                    )
                }
                .item()
                .model { c: DataGenContext<Item?, BlockItem?>, p: RegistrateItemModelProvider ->
                    p.withExistingParent(c.name, baseItemModelLocation)
                        .texture("0", baseBlockModelLocation)
                        .texture("1", coreTextureLocation)
                        .texture("2", liquidTextureLocation)
                }
                .build()
        }
    }

    fun <B : Block?, P> flapbearing(): NonNullUnaryOperator<BlockBuilder<B, P>> {
        val baseBlockModelLocation: ResourceLocation = ClockworkMod.asResource("block/flap_bearing/block")
        val baseItemModelLocation: ResourceLocation = ClockworkMod.asResource("block/bearing/item")
        val topTextureLocation: ResourceLocation = ClockworkMod.asResource("block/flap_bearing/top")
        val baseTextureLocation: ResourceLocation = ClockworkMod.asResource("block/flap_bearing")
        return NonNullUnaryOperator { b: BlockBuilder<B, P> ->
            b.initialProperties { SharedProperties.stone() }
                .properties { p: BlockBehaviour.Properties -> p.noOcclusion() }
                .blockstate { c: DataGenContext<Block?, B>, p: RegistrateBlockstateProvider ->
                    p.directionalBlock(
                        c.get(), p.models()
                            .withExistingParent(c.name, baseBlockModelLocation)
                            .texture("0", baseBlockModelLocation)
                            .texture("1", topTextureLocation)
                            .texture("2", baseTextureLocation)
                    )
                }
                .item()
                .model { c: DataGenContext<Item?, BlockItem?>, p: RegistrateItemModelProvider ->
                    p.withExistingParent(c.name, baseItemModelLocation)
                        .texture("0", baseBlockModelLocation)
                        .texture("1", topTextureLocation)
                        .texture("2", baseTextureLocation)
                }
                .build()
        }
    }

    fun <B : Block?, P> bearing(
        prefix: String?,
        backTexture: String?
    ): NonNullUnaryOperator<BlockBuilder<B, P>> {
        val baseBlockModelLocation = ClockworkMod.asResource("block/bearing/block")
        val baseItemModelLocation = ClockworkMod.asResource("block/bearing/item")
        val topTextureLocation = ClockworkMod.asResource("block/bearing_top")
        val sideTextureLocation = ClockworkMod.asResource("block/" + prefix + "_bearing_side")
        val backTextureLocation = ClockworkMod.asResource("block/" + backTexture)
        return NonNullUnaryOperator { b: BlockBuilder<B, P>? ->
            b!!.initialProperties(NonNullSupplier { SharedProperties.stone() })
                .properties(NonNullUnaryOperator { p: BlockBehaviour.Properties -> p.noOcclusion() })
                .blockstate(NonNullBiConsumer { c: DataGenContext<Block, B>, p: RegistrateBlockstateProvider? ->
                    p!!.directionalBlock(
                        c!!.get(), p.models()
                            .withExistingParent(c.getName(), baseBlockModelLocation)
                            .texture("side", sideTextureLocation)
                            .texture("back", backTextureLocation)
                    )
                })
                .item()
                .model(NonNullBiConsumer { c: DataGenContext<Item?, BlockItem?>?, p: RegistrateItemModelProvider? ->
                    p!!.withExistingParent(c!!.getName(), baseItemModelLocation)
                        .texture("top", topTextureLocation)
                        .texture("side", sideTextureLocation)
                        .texture("back", backTextureLocation)
                })
                .build()
        }
    }
}
